src/theory/datatypes/infer_proof_cons.cpp

```diff
@@ -317,9 +317,50 @@ void InferProofCons::convert(InferenceId infer, TNode conc, TNode exp, CDProof*
       }
     }
     break;
+    case InferenceId::DATATYPES_CYCLE:
+    {
+      // the conflict is of the form
+      // (and (= x (C1 ... x1 ...))
+      //      (= x1 (C2 ... x2 ...)) ....
+      //      (= x{n-1} (C2 ... xn ...))
+      //      (= xn (Cn ... x ...)))
+      // We take the first n-1 equalities as a substitution and apply it to
+      // the right hand side of the last equality, and use DT_CYCLE to derive
+      // a conflict.
+      Assert(!expv.empty());
+      Node lastEq = expv[expv.size() - 1];
+      Assert(lastEq.getKind() == Kind::EQUAL);
+      std::vector<Node> subs(expv.begin(), expv.begin() + expv.size() - 1);
+      ProofChecker* pc = d_env.getProofNodeManager()->getChecker();
+      Node eq;
+      if (!subs.empty())
+      {
+        eq = pc->checkDebug(ProofRule::SUBS, subs, {lastEq[1]});
+        Assert(!eq.isNull());
+        cdp->addStep(eq, ProofRule::SUBS, subs, {lastEq[1]});
+      }
+      else
+      {
+        eq = lastEq[1].eqNode(lastEq[1]);
+      }
+      Node eq1 = lastEq[0].eqNode(eq[1]);
+      Trace("dt-ipc-cycle") << "Cycle eq? " << eq1 << std::endl;
+      Node falsen =
+          d_env.getRewriter()->rewriteViaRule(ProofRewriteRule::DT_CYCLE, eq1);
+      if (!falsen.isNull())
```

Comment:
shouldn't you be setting `success` inside this block?

Reply:
Good catch, apparently it didnt matter since we don't overwrite proof steps, but indeed it should be set.